File: Stacks and queues/MinStack.java
import java.util.Scanner;
import java.util.Stack;

public class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    public void pop() {
        if (stack.isEmpty()) {
            System.out.println("Stack is empty!");
            return;
        }
        if (stack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        stack.pop();
    }

    public int top() {
        if (stack.isEmpty()) {
            System.out.println("Stack is empty!");
            return -1;
        }
        return stack.peek();
    }

    public int getMin() {
        if (minStack.isEmpty()) {
            System.out.println("Stack is empty!");
            return -1;
        }
        return minStack.peek();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        MinStack minStack = new MinStack();

        while (true) {
            System.out.println("Enter command: push X / pop / top / getMin / exit");
            String input = sc.nextLine().trim();
            if (input.equalsIgnoreCase("exit")) break;

            if (input.startsWith("push")) {
                String[] parts = input.split(" ");
                int val = Integer.parseInt(parts[1]);
                minStack.push(val);
                System.out.println(val + " pushed");
            } else if (input.equalsIgnoreCase("pop")) {
                minStack.pop();
            } else if (input.equalsIgnoreCase("top")) {
                System.out.println("Top: " + minStack.top());
            } else if (input.equalsIgnoreCase("getMin")) {
                System.out.println("Min: " + minStack.getMin());
            } else {
                System.out.println("Invalid command!");
            }
        }
        sc.close();
        System.out.println("Program terminated.");
    }
}

File: Stacks and queues/MinStack.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

class MinStack {
    stack<int> st;
    stack<int> minSt;

public:
    void push(int val) {
        st.push(val);
        if (minSt.empty() || val <= minSt.top()) {
            minSt.push(val);
        }
        cout << val << " pushed\n";
    }

    void pop() {
        if (st.empty()) {
            cout << "Stack is empty!\n";
            return;
        }
        if (st.top() == minSt.top()) {
            minSt.pop();
        }
        cout << st.top() << " popped\n";
        st.pop();
    }

    void top() {
        if (st.empty()) {
            cout << "Stack is empty!\n";
            return;
        }
        cout << "Top: " << st.top() << "\n";
    }

    void getMin() {
        if (minSt.empty()) {
            cout << "Stack is empty!\n";
            return;
        }
        cout << "Min: " << minSt.top() << "\n";
    }
};

int main() {
    MinStack minStack;
    string command;

    cout << "Enter commands: push X / pop / top / getMin / exit\n";

    while (true) {
        cout << "> ";
        getline(cin, command);

        if (command == "exit") break;

        if (command.substr(0, 4) == "push") {
            int val = stoi(command.substr(5));
            minStack.push(val);
        } else if (command == "pop") {
            minStack.pop();
        } else if (command == "top") {
            minStack.top();
        } else if (command == "getMin") {
            minStack.getMin();
        } else {
            cout << "Invalid command!\n";
        }
    }

    cout << "Program terminated.\n";
    return 0;
}
